### ISSUE TITLE
Code-Unification of CUDA runtime and OpenCL driver

### ISSUE THREAD

**Issue by SeanSiders** [G-PIES/G-PIES#20]
At the very least we could have a script which generates the .cl and .cu files prior to compiling, for the CD and OKMC code. Either way we need to avoid code duplication in the development pipeline.

**Comment by SeanSiders** [comment]
There are existing examples of OpenCL/CUDA unification in the molecular dynamics libraries LAMMPS and OpenMM. In a [discussion amongst OpenMM developers](https://github.com/openmm/openmm/issues/2471) they mention AMD's compiler [ROCm](https://github.com/ROCm/ROCm) as a potential solution. LAMMPS has a proprietary solution, [geryon](https://github.com/lammps/lammps/tree/develop/lib/gpu/geryon) which provides a common computing interface for their GPU kernels (under the BSD license).

**TL;DR**
The ROCm compiler could be a great solution for us BUT if we want to consider long-term support for MacOS, we will also need to consider their GPU API Metal _shrug_... The work to implement a common interface is simply too much for now but these resources will prove useful if we consider it in the future.
